ksmt-core/src/main/kotlin/org/ksmt/expr/KExpr.kt

```diff
@@ -7,7 +7,7 @@ import org.ksmt.expr.transformer.KTransformerBase
 import org.ksmt.sort.KSort
 import org.ksmt.expr.printer.ExpressionPrinterWithLetBindings
 
-abstract class KExpr<T : KSort>(ctx: KContext) : KAst(ctx) {
+abstract class KExpr<out T : KSort>(ctx: KContext) : KAst(ctx) {
```

Comment:
A while ago, we had this `out` (I don't remember in a prototype or in the main branch) but removed it due to some now possible assignments with the wrong types. I don't remember details at all, but won't it cause issues with strict generic types from some caches?

Reply:
Previous attempt to introduce `out` was in #13. 
We first add `out` in this [commit](https://github.com/UnitTestBot/ksmt/pull/13/commits/f9a549081e28333c79c30b6ffe1514b9c134233c) and then drop it in this [commit](https://github.com/UnitTestBot/ksmt/commit/d87ff884c019e2f077c34df1f958d67fd25e154e). 

We decided to remove `out` after some offline conversation where we mentioned some "problems". If i remember correctly, there were some problems with `KArith*`, because of tricky sort type constraints. 

I don't see such problems right now, but maybe i'm missing something.

Reply:
The problem is: 
```
val x: KExpr<KIntSort> 
val y: KExpr<KBvSort> 
val e = mkIte(trueExpr, x, y)
```